qwen_agent/llm/openvino.py

```diff
@@ -0,0 +1,154 @@
+from transformers import (AutoTokenizer, AutoConfig,
+                          TextIteratorStreamer, StoppingCriteriaList, StoppingCriteria)
+
+from qwen_agent.llm.base import register_llm
+from qwen_agent.llm.schema import ASSISTANT, DEFAULT_SYSTEM_MESSAGE, SYSTEM, USER, Message
+from qwen_agent.llm.text_base import BaseTextChatModel
+from qwen_agent.log import logger
+
+
+from transformers.generation.stopping_criteria import StoppingCriteriaList, StoppingCriteria
+
+
+class StopSequenceCriteria(StoppingCriteria):
+    """
+    This class can be used to stop generation whenever a sequence of tokens is encountered.
+
+    Args:
+        stop_sequences (`str` or `List[str]`):
+            The sequence (or list of sequences) on which to stop execution.
+        tokenizer:
+            The tokenizer used to decode the model outputs.
+    """
+
+    def __init__(self, stop_sequences, tokenizer):
+        if isinstance(stop_sequences, str):
+            stop_sequences = [stop_sequences]
+        self.stop_sequences = stop_sequences
+        self.tokenizer = tokenizer
+
+    def __call__(self, input_ids, scores, **kwargs) -> bool:
+        decoded_output = self.tokenizer.decode(input_ids.tolist()[0])
+        return any(decoded_output.endswith(stop_sequence) for stop_sequence in self.stop_sequences)
+
+
+@register_llm('openvino')
+class OpenVINO(BaseTextChatModel):
+    """OpenVINO Pipeline API.
+
+    To use, you should have the ``optimum[openvino,nncf]`` python package installed.
+
+    Example export and quantizie from_model_id by command line:
+        optimum-cli export openvino --model Qwen/Qwen2-7B-Instruct --task text-generation-with-past --weight-format int4 --group-size 128 --ratio 0.8 Qwen2-7B-Instruct-ov
+    
+    Example passing pipeline in directly:
+        llm_cfg = {
+            'ov_model_dir': 'Qwen2-7B-Instruct-ov',
+            'model_type': 'openvino'
+            }
+        system_instruction = '''You are a helpful assistant.
+        After receiving the user's request, you should:
+        - first draw an image and obtain the image url,
+        - then run code `request.get(image_url)` to download the image,
+        - and finally select an image operation from the given document to process the image.
+        Please show the image using `plt.show()`.'''
+        tools = ['my_image_gen', 'code_interpreter']
+        files = ['./examples/resource/doc.pdf']
+        bot = Assistant(llm=llm_cfg,
+                system_message=system_instruction,
+                function_list=tools,
+                files=files)
+    """
+
+    def __init__(self, cfg: Optional[Dict] = None):
+        super().__init__(cfg)
+        if 'ov_model_dir' not in cfg:
+            raise ValueError(
+                f'Please provide openvino model directory through `ov_model_dir` in cfg')
+
+        self.ov_model = OVModelForCausalLM.from_pretrained(
+            cfg['ov_model_dir'],
+            device=cfg.get('device', 'cpu'),
```

Comment:
does this device parameter in config used somewhere else  outside  pipeline (e.g. for some inputs/outputs preparing as pytorch model device)?
if yes, there we can have some possible confusion between gpu vs cuda. Could you please check that it works for openvino  gpu

Reply:
hi, as my test, it works well with `device="GPU"`